Ingest malops, sensors, malwares, remediations

Dhaiwat10 opened the pull request.

Dhaiwat10 commented on src/steps/malwares/converters.ts:
```diff
@@ -0,0 +1,24 @@
+        _key: getMalwareKey(data.guid),
+        category: 'malware',
+        open: data.status === 'Detected' ? true : false,
+        severity: data.score.toString(), // doubtful
+        numericSeverity: data.score,
```
The API only provides the severity in terms of numerics (in the form of the field score). Not sure how we should handle converting them to 'strings'.

eXtremeX replied:
Let's tackle this once we're done with "big pieces" first. We could try of thinking about some conversion, but if we fail to come up with a solution we think make perfect sense, we can always ask JupiterOne people as well!

Dhaiwat10 commented on src/steps/malwares/index.ts:
```diff
@@ -0,0 +1,29 @@
+import { Entities, IntegrationSteps } from '../constants';
+import { createMalwareEntity } from './converters';
+
+export async function fetchMalwares({
+  instance,
+  jobState,
+}: IntegrationStepExecutionContext<IntegrationConfig>) {
+  const apiClient = createAPIClient(instance.config);
+  await apiClient.iterateMalwares(async (malware) => {
+    const malwareEntity = createMalwareEntity(malware);
+    await jobState.addEntity(malwareEntity);
+  });
+}
+
```
This step throws a DuplicateKey error. You will see it when you run yarn start. Apparently the API is sending back duplicate records.

Dhaiwat10 commented on src/types.ts:
```diff
@@ -0,0 +1,152 @@
+  elementValues: {
+    affectedUsers: ElementMalopFeature;
+    affectedMachines: ElementMalopFeature;
+  };
+  suspicions: null; // Unclear
+  filterData: {
+    sortInGroupValue: string;
```
Have added // Unclear wherever the type of the respective field was not clear from the API. (In most cases, these fields were marked as null.)

Dhaiwat10 commented on docs/jupiterone.md:
```diff
@@ -105,6 +113,7 @@ The following relationships are created/mapped:
 | `acme_account`        | **HAS**               | `acme_group`          |
 | `acme_account`        | **HAS**               | `acme_user`           |
 | `acme_group`          | **HAS**               | `acme_user`           |
+| `malop`               | **HAS**               | `remediation`         |
 
```
Was able to add one relationship.

Dhaiwat10 commented on src/client.ts:
```diff
@@ -207,6 +212,10 @@ export class APIClient {
 
     const malwares = body.data.malwares;
 
+    // TODO: Only keep the latest if there are duplicates
+    // If there is a duplicate of a malware and they share the same guid and machineName
+    // Only keep the one that has the highest timestamp and don't iterate over the others
+
```
This makes sense but I am worried that this would require lots of iteration over the data that we receive. I can see this affecting the _'speed'_ drastically. Is that a concern for us?

eXtremeX replied:
I'm afraid this is necessary, the reason being is the graph data is supposed to represent the current snapshot of all the resources/entities, it shouldn't hold some historical data in it, but this is not a problem that we're often faced with. Whatever we decide to do we'll check back with JupiterOne later and see if it requires further changes.

Options that we have currently are:
1) Make timestamp be part of the key. This would result in no duplicates, but if we ever later need to find some malware via its key, it might be impossible (e.g. we might know the guid of malware but not the timestamp)
2) The filtering idea I added in the comments.

In my opinion, just go with that filtering option for and we'll see if it'll require some changes later down the road.

Dhaiwat10 commented on src/steps/sensors/converters.ts:
```diff
@@ -16,6 +16,8 @@ export function createSensorEntity(data: Sensor) {
         _key: getSensorKey(data.guid),
         name: data.guid,
         category: ['sensor'],
+        // TODO: can you map all the fields in here?
+        // Similar to malops/converter.ts (just you don't have use the custom map function)
```
Sure. Just for the sensors, or for all the remaining entities as well?

eXtremeX replied:
For the remaining entities as well - we want to map over most/if not all fields inside of our converters :slightly_smiling_face:

Dhaiwat10 commented on src/steps/malops/converters.ts:
```diff
@@ -14,8 +35,85 @@ export function createMalopEntity(data: Malop) {
+            )
+          : undefined,
+        isBlocked: mapMalopSimpleFieldValue(data.simpleValues.isBlocked),
+        rootCauseElementTypes: mapMalopSimpleFieldValue(
+          data.simpleValues.rootCauseElementTypes,
+        ),
+        rootCauseElementNames: mapMalopSimpleFieldValue(
+          data.simpleValues.rootCauseElementNames,
+        ),
+        malopLastUpdateTime: mapMalopSimpleFieldValue(
+          data.simpleValues.malopLastUpdateTime,
+        ),
+        allRansomwareProcessesSuspended: mapMalopSimpleFieldValue(
+          data.simpleValues.allRansomwareProcessesSuspended,
+        ),
+        managementStatus: mapMalopSimpleFieldValue(
+          data.simpleValues.managementStatus,
+        ),
+        closeTime: mapMalopSimpleFieldValue(data.simpleValues.closeTime)
+          ? parseInt(
+              mapMalopSimpleFieldValue(data.simpleValues.closeTime) as string,
+              10,
+            )
+          : undefined,
+        // In general, whenever we know some value can be null, we'd want to set it as undefined instead
+        closerName:
+          mapMalopSimpleFieldValue(data.simpleValues.closerName) || undefined,
+        // elementValues seem useful however they're not primitive types and neither arrays of primitive types
+        // which means we couldn't just map them like above, it'd be best if we leave them for later and check with J1
+        // sometimes this requires us to create new entities (that don't map to any of the endpoint API calls) but
+        // that we can connect with this entity
+        suspicions: data.suspicions || undefined,
+        suspicionCount: data.suspicionCount,
+        isMalicious: data.isMalicious,
+        malopPriority: data.malopPriority || undefined,
+        // Required fields for class "Assessment"
+        category: 'Risk Assessment',
+        summary:
+          data.isMalicious === false ? 'Is not malicious' : 'Is malicious',
+        internal: true,
```
Is this something we do for all entities usually (referring to the 'fields' you added here)? Or is this something specific for this case? A bit confused since I went through the WPEngine integration that Paul worked on and didn't come across such a thing.

eXtremeX replied:
Are you referring to the fact that I had to create that `mapMalopSimpleFieldValue` function or the fact that I introduced all the fields?

If former - if the mapping can't be made directly (mapping can be made directly if the target field is a primitive type or an array of primitive types), we need to resort to something like this.

If latter - yes we usually want most of the fields to land on our entities. Granted if we're sure that some field doesn't make sense or isn't relevant or isn't necessary - we don't want to port it over, though it's usually hard to decide if something truly is useless (there are a lot of customers and a lot of use-cases) so going with "broad strokes" initially is the safest approach.

Dhaiwat10 commented on src/steps/malops/converters.ts:
```diff
@@ -1,11 +1,32 @@
   return `malop:${id}`;
 }
 
+// TODO: some improvements:
+// 1) If the value is boolean in string form - e.g. "false" or "true", convert it to boolean
```
Tried the following fix. I was able to convert all of them to booleans but it causes a J1 Graph/schema error since we have defined these properties as of type `string` and not `boolean`. We could easily change their types in the types file, but wouldn't that cause the fetchMalops step to throw an error? (because they would be of type `string` originally)

```javascript
function toBool(val: any) {
  if (val === 'true') {
    return true;
  }

  if (val === 'false') {
    return false;
  }

  return val;
}

function mapMalopSimpleFieldValue(
  value: SimpleMalopFeature,
): boolean | string | string[] | null {
  if (!value) {
    return null;
  }
  // We know the return type isn't going to be an array
  if (value.totalValues && value.totalValues === 1) {
    if (value.values && value.values.length > 0) {
      return toBool(value.values[0]);
    }
  }

  return value.values ? value.values.map((val) => toBool(val)) : null;
}
```

Dhaiwat10 replied:
Tried the following fix. I was able to convert all of them to booleans but it causes a J1 Graph/schema error since we have defined this properties as of type `string` and not `boolean`. We could easily change their types in the types file, but wouldn't that cause the fetchMalops step to throw an error? (because they would be of type `string` originally)

```javascript
function toBool(val: any) {
  if (val === 'true') {
    return true;
  }

  if (val === 'false') {
    return false;
  }

  return val;
}

function mapMalopSimpleFieldValue(
  value: SimpleMalopFeature,
): boolean | string | string[] | null {
  if (!value) {
    return null;
  }
  // We know the return type isn't going to be an array
  if (value.totalValues && value.totalValues === 1) {
    if (value.values && value.values.length > 0) {
      return toBool(value.values[0]);
    }
  }

  return value.values ? value.values.map((val) => toBool(val)) : null;
}
```

Dhaiwat10 commented:
Fixed the duplicate malwares bug. Added the mappings like you asked in the sensors converter. Cannot push anything because the tests are failing.

The malops test fails because of the issue with the types caused by the boolean conversion. (please see the comment above, will need your opinion on this one)

The sensors test fails because of `Unexpected JSON response at https:xxxxx/login.html` inside the `iterateSensors` function. Currently trying to debug this myself. 

Meanwhile I am working on writing tests & creating relationships for the rest of the entities.
